nixos/modules/services/network-filesystems/ipfs.nix

```diff
@@ -0,0 +1,118 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  inherit (pkgs) ipfs;
+
+  cfg = config.services.ipfs;
+
+
+  ipfsFlags = ''${if cfg.autoMigrate then "--migrate" else ""}
+                ${if cfg.enableGC    then "--enable-gc" else ""}
+                ${toString cfg.extraFlags}'';
+
+in
+
+{
+
+  ###### interface
+
+  options = {
+
+    services.ipfs = {
+
+      enable = mkOption {
+        default = false;
+        description = ''
+          Whether to start a daemon for the Interplanetary
+          file system (IPFS).
+        '';
+      };
+
+      user = mkOption {
+        type = types.str;
+        default = "ipfs";
+        description = "User under which the IPFS daemon runs";
+      };
+
+      group = mkOption {
+        type = types.str;
+        default = "ipfs";
+        description = "Group under which the IPFS daemon runs";
+      };
+
+      dataDir = mkOption {
+        type = types.str;
+        default = "/var/lib/ipfs";
+        description = "The data dir for IPFS";
+      };
+
+      autoMigrate = mkOption {
+        default = false;
+        description = ''
+          Whether IPFS should try to migrate the file system automatically.
+        '';
+      };
+
+      enableGC = mkOption {
+        default = false;
+        description = ''
+          Whether to enable automatic garbadge collection.
+        '';
+      };
+
+      extraFlags = mkOption {
+        type = types.listOf types.str;
+        description = "Extra flags passed to the IPFS daemon";
+        default = [];
+      };
+    };
+  };
+
+  ###### implementation
+
+  config = mkIf cfg.enable {
+    environment.systemPackages = [ pkgs.ipfs ];
+
+    users.extraUsers = optionalAttrs (cfg.user == "ipfs") singleton {
+      name = "ipfs";
+      group = cfg.group;
+      home = cfg.dataDir;
+      createHome = false;
+      uid = config.ids.uids.ipfs;
+      description = "IPFS daemon user";
+    };
+
+    users.extraGroups = optionalAttrs (cfg.group == "ipfs") singleton {
+      name = "ipfs";
+      gid = config.ids.gids.ipfs;
+    };
+
+    systemd.services.ipfs = {
+      description = "IPFS Daemon";
+
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network.target" "local-fs.target" ];
+      path  = [ pkgs.ipfs pkgs.runit ];
+
+      preStart =
+        ''
+          mkdir -m 0755 -p ${cfg.dataDir}
+          chown ${cfg.user}:${cfg.group} ${cfg.dataDir}
+          if [[ ! -d ${cfg.dataDir}/.ipfs ]]; then
+            cd ${cfg.dataDir}
+            ${pkgs.runit}/bin/chpst -u "${cfg.user}:${cfg.group}" ipfs init
```

Comment:
Sure. Changed it accordingly. runit's `chpst` is just a nice alternative if the user does not have a login shell since you need to provide that with `su`.